decisions/0001-bundle-granularity.rst

```diff
@@ -0,0 +1,199 @@
+model our courses with OLX data.
+
+One late change to our overall philosophical outlook was to radically separate
+the human readable format used in import/export from the machine-readable format
+that provides organization and reuse capability in Blockstore itself. This
+implies shifting more of the higher level OLX -> Blockstore mapping knowledge
+to be outside of Blockstore itself, and gives us opportunities to make some
+simplifications.
+
+---------------------------------------------------
+Proposal: OLX Bundles with Flat Directory Structure
+---------------------------------------------------
+
+In this proposal, a Bundle for any chunk of OLX content has top-level folders
+for every definition in the Bundle, including both leaf and container nodes. A
+problem in a problem bank may have just one such directory for the problem
+itself, or a few directories representing a Unit and multiple leaf nodes. A
+course may potentially have thousands of directories.
+
+The major focus of this approach is operational simplicity at scale. It is
+generally less featureful than other proposals and shifts more of the work to
+OLX-aware clients.
+
+An important mental shift is that this walks back from some of the file system
+analogies of the original proposal. We still have files in Bundles, and XBlock
+content is still stored in directories with their associated static assets. But
+when we make data references across XBlocks, it's by identifiers and conventions
+and not by relative file paths. It's the job of clients to transform that into
+a more friendly file-path based format for the human readable format.
+
+XBlock Directories
+==================
+
+XBlock Directories are a client convention and not a core Blockstore primitive.
+Each directory represents a single XBlock and its associated assets. The naming
+scheme is::
+
+    /{definition-key}/
+                     {tag-type}.xml
+                     static/
+
+The ``static`` directory is optional. For a concrete example::
+
+    /problem+blockstore_granularity/
+                                    problem.xml
+                                    static/
+                                           diagram.png
+
+Meanwhile, a container looks like:
+
+    /unit+blockstore_big_questions/problem.xml
+
+When containers reference their children, it's done via a new tag. So a
+particular ``unit.xml`` might look like::
+
+    <unit display_name="">
+        <!-- Normal usage -->
+        <xblock-include definition="unit+blockstore_big_questions" />
+
+        <!-- If we want to include the same content with a modifier to the usage
+             key so that we can include it multiple times with different state
+             storage. -->
+        <xblock-include definition="unit+blockstore_big_questions" usage="alternate" />
+
+        <!-- XBlock that exists in a Linked Bundle aliased to "arch_hangout_vidoes" -->
+        <xblock-include definition="arch_hangout_videos/video+blockstore_overview"/>
```

Comment:
I am assuming the definition keys in a single bundle are going to be unique. But can one context refer to two XBlocks from two different bundles which have the same definition key?

Reply:
> can one context refer to two XBlocks from two different bundles which have the same definition key?

I don't see why not? Specifying a unique `usage` key for each would likely be required in that case though:

```
<xblock-include definition="external_bundle_1/video+blockstore_overview" usage="vid1"/>
<xblock-include definition="external_bundle_2/video+blockstore_overview" usage="vid2"/>
```

Reply:
I just want to confirm I understand the usage key properly. They'll need to to be unique, right? If that's the case, could we change the example to `usage="alternate1"` to emphasize this?

Reply:
> If they are two different XBlocks, should there not be two different ids for them?

They have different definition IDs (some combo of source bundle UUID + folder name in source bundle) and they have different usage keys here (CourseKey is same but definition and usage are each different). So in fact there are no identifiers that are the same.

In other words, `external_bundle_1/video+blockstore_overview` and `external_bundle_2/video+blockstore_overview` are different IDs, and so are "vid1" and "vid2"

Or in other words, we should always think of definition keys as including both the bundle UUID and the human-provided slug.